### PR TITLE
New instanced rendering example

### PR DESCRIPTION
**Pull request by brisvag**
Turns out, instanced rendering works out of the box in vispy!

Here's a modified version of `examples/tutorial/gl/quad.py` which uses it to render a bunch of quads.

For those like me who know nothing, the changes needed are:

- you have to request the full `gl+` context
- you need to create/bind a new buffer containing whatever attributes need to change on a per-instance basis
- you have to set the `glVertexAttribDivisor` for each attribute that needs to change on an per-instance basis, to the number of instances should be rendered before going to the next elements... in practice, 1 for most applications.
- you have to call `glDrawArraysInstanced` instead of `glDrawArrays` (or the corresponding `Element` draw call)

Now, this example is all well and good, but it would be *awesome* if I could trickle this up to the `visual` world. I would appreciate tips on where this might live. My best guess is some changes are needed here:

https://github.com/vispy/vispy/blob/6b9d39c8c7ab75ceb7f15ccccef5f0d27eaeeec1/vispy/gloo/glir.py#L1332-L1350

and we need to somehow:
- tell a vertex buffer that it should set a `Divisor`
- switch to the correct draw call if any of the bound vertex buffers have a `divisor`

Does any of this make sense?

### CODE REVIEW

**Comment by kephale** [comment]
@brisvag It is working on my M1:

image

**Comment by djhoese** [comment]
This is pretty cool. I don't really know much about it but it is neat that you got it working.

The build website job is failing because of pydata sphinx theme. We could limit the version in the CI job. Otherwise we need to just wait for them to fix the issue. I commented on the issue I had made to ask about a bug fix release.

Feel free to merge. However, you mention the other `quad.py` example. Do we want to move all the examples in that directory to the `gloo` gallery directory and then use sphinx references to link to the other example from within your docstring in your new example?

**Comment by brisvag** [comment]
> Do we want to move all the examples in that directory to the `gloo` gallery directory and then use sphinx references to link to the other example from within your docstring in your new example?

Yup, I can do that! Can you point me to some example fo how this works? Not sure how to do it from a docstring.

**Comment by djhoese** [comment]
According to the sphinx-gallery docs, it should be generating a sphinx ref that you can link to:

https://sphinx-gallery.github.io/stable/advanced.html#cross-referencing

I thought I was doing it for other examples or in the vispy docs but a `grep` locally isn't showing anything. So you could maybe do:

```
Based on the :ref:`original quad example <sphx_glr_gallery_gloo_quad.py>`.
```

Or something like that.

**Comment by brisvag** [comment]
Based on your #2378, cross references won't help here, since neither example is rendered in the gallery, right?

**Comment by djhoese** [comment]
Yes, that's why I said:

> Do we want to move all the examples in that directory to the gloo gallery directory

So that they would be rendered with the gallery.

**Comment by brisvag** [comment]
Oops, misunderstood :) Should we keep the `gl`/`gloo` separation and add a new gallery, given that these examples are much lower level compared to the gloo machinery?

**Comment by djhoese** [comment]
If we're going to add another gallery it shouldn't be a low-level one. VisPy 2.0 won't be specific to OpenGL at all and I don't want to advertise the functionality to users as a "you should use this" thing. No one should use the really low-level stuff unless they are contributing directly to vispy. How about a sub-directory in the `gloo` gallery for low-level `gl` examples? I *think* you can put a README (.rst or .md, not sure which) that describes what the sub-directory is about. I did that for my realtime data tutorial in the scene gallery.

**Comment by brisvag** [comment]
Done!

File: examples/gloo/gl/README.rst [hunk 1]
```diff
@@ -0,0 +1,4 @@
+The low-level GL module
+=======================
+
+Vispy has a low-level `gl` module which wraps `GL` calls almost to a 1-to-1 ratio. While this intended mostly for internal use, it is possible to use this to construct programs.
```
Comment (djhoese) [inline examples/gloo/gl/README.rst:4]:
```suggestion
Vispy has a low-level `gl` module which wraps `GL` calls almost to a 1-to-1 ratio. While this is intended mostly for internal use, it is possible to use this to construct programs.
```

**Comment by brisvag** [comment]
@djhoese the website error is unrelated, right? In which case, is it ok to merge this?

**Comment by djhoese** [comment]
Since this PR updates an example, let's just force the version of pydata-sphinx-theme here:

https://github.com/vispy/vispy/blob/16d8c5bd8694df9c18086e6fad31ffbaa08abaf4/ci/requirements/linux_website_deps_pip.txt#L9

Can you set this to `!= 0.10` and see if that fixes it? There is a 0.10.0 and a 0.10.1 both which include a bug that makes our docs fail. I'm not sure the syntax for a requirements file like this. Maybe a space then the specifier?

File: ci/requirements/linux_website_deps_pip.txt [hunk 2]
```diff
@@ -6,7 +6,7 @@
  sphinx_gallery
  sphinxcontrib-apidoc
  myst-parser
- pydata-sphinx-theme
+ pydata-sphinx-theme!=0.10
```
Comment (djhoese) [inline ci/requirements/linux_website_deps_pip.txt:9]: Looks like this still matches against 0.10.1:

```
 Collecting pydata-sphinx-theme!=0.10
  Downloading pydata_sphinx_theme-0.10.1-py3-none-any.whl (1.5 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.5/1.5 MB 129.4 MB/s eta 0:00:00
```

Any idea what magic comparison operator has to be used to not get 0.10.x?

Comment (brisvag) [reply ci/requirements/linux_website_deps_pip.txt:9]: But that won't prevent ending up with 0.10 if other constraints in their enviroment cause it.

**Comment by brisvag** [comment]
Still fialing, but I can't figure out if it's the same issue or something else... Where do you find the installed library in the log?

**Comment by djhoese** [comment]
New error:

```
 /home/runner/work/vispy/vispy/doc/gallery/gloo/index.rst:395: WARNING: toctree contains reference to document 'gallery/gloo/texture_precision' that doesn't have a title: no link will be generated
```


As for parsing out the log, on this page:

https://github.com/vispy/vispy/actions/runs/3157145392/jobs/5137658853

You have two install sections/jobs, one for conda, one for pip. This particular one is a pip dependency:

```
 Collecting pydata-sphinx-theme!=0.10.*
  Downloading pydata_sphinx_theme-0.9.0-py3-none-any.whl (3.3 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.3/3.3 MB 49.3 MB/s eta 0:00:00
```

**Comment by brisvag** [comment]
Everything seems to be doing ok, merging!

**Pull request merged**